### ISSUE TITLE
ExtendedErrorInformation is null when HTTP body is present

### ISSUE THREAD

**Issue by davidmatson** [Azure/azure-storage-net#271]
Recently, ExtendedErrorInformation on StorageException started being null some of the time when it should not be. This is using SDK version 4.3.0, but I believe the code is unchanged in the latest. (Let me know if that's not the case.)

The following code (simplified repro) should work, but sometimes does not:

``` c#
try
{
    blob.Delete();
}
catch (StorageException ex)
{
    if (ex.RequestInformation.ExtendedErrorInformation != null && ex.RequestInformation.ExtendedErrorInformation.ErrorCode = "BlobNotFound")
    {
         // Expected possibility; continue
    }
    else
    {
        throw;
    }
}
```

The code above usually continues but recently started throwing when it should not. Examining the exception, the ExtendedErrorInformation returns null. But when viewing the inner exception, I'm able to get valid ExtendedErrorInformation back. Using the following code:

``` c#
StorageExtendedErrorInformation extendedInformation = ex.RequestInformation.ExtendedErrorInformation;

if (extendedInformation == null)
{
    Console.WriteLine("Null extended information");
}
else
{
    Console.WriteLine("ErrorCode: {0}", extendedInformation.ErrorCode);
    Console.WriteLine("ErrorMessage: {0}", extendedInformation.ErrorMessage);
}

WebException webException = (WebException)ex.InnerException;
HttpWebResponse response = (HttpWebResponse)webException.Response;
MemoryStream s = (MemoryStream)response.GetResponseStream();
MemoryStream s2 = new MemoryStream(s.GetBuffer(), 0, (int)response.ContentLength);
TextReader r = new StreamReader(s2);
Console.WriteLine(r.ReadToEnd());
s2.Position = 0;
extendedInformation = StorageExtendedErrorInformation.ReadFromStream(s2);

if (extendedInformation == null)
{
    Console.WriteLine("Null extended information");
}
else
{
    Console.WriteLine("ErrorCode: {0}", extendedInformation.ErrorCode);
    Console.WriteLine("ErrorMessage: {0}", extendedInformation.ErrorMessage);
}
```

it outputs:

> Null extended information
> Content: <?xml version="1.0" encoding="utf-8"?><Error><Code>BlobNotFound</Code><Message>The specified blob does not exist.
> RequestId:37af06c8-0001-0106-4fd0-a4815b000000
> Time:2016-05-03T00:12:08.3285784Z</Message></Error>
> ErrorCode: BlobNotFound
> ErrorMessage: The specified blob does not exist.
> RequestId:37af06c8-0001-0106-4fd0-a4815b000000
> Time:2016-05-03T00:12:08.3285784Z

In other words, the HTTP Response message actually has an XML body that contains the extended error information in a valid format that can be used if the response buffer stream is accessed directly, but when StorageException is constructed, the ExtendedErrorInformation is set to null.

Note that this problem does not occur on every attempt, on some, so I suspect there's something in the interaction between StorageException and HttpWebResponse's stream that is causing the problem.

**Comment by davidmatson** [comment]
I found the root cause - we didn't have a binding redirect to Microsoft.Data.Services.Client.dll. (I believe the same thing would have happened if this DLL were missing at runtime.)

Ironically, the line where we're trying to load this DLL says it should never be used anyway (StorageException.cs):
> 350                 // Should never get to this one since we will call TranslateDataServiceException for DataService operations. 
> 351                 StorageException tableEx = TableUtilities.TranslateDataServiceException(ex, reqResult, null); 
> 352 
 
> 353                 if (tableEx != null) 
> 354                 { 
> 355                     return tableEx; 
> 356                 } 

But the real bug is in TranslateException, which just swallows all exceptions and returns a StorageExceptino with incomplete data (!)
> 156             try 
> 157             { 
> 158                 if ((storageException = CoreTranslate(ex, reqResult, ref parseError)) != null) 
> 159                 { 
> 160                     return storageException; 
> 161                 } 
> ...
> 175             } 
> 176             catch (Exception) 
> 177             { 
>178                 // if there is an error thrown while parsing the service error, just wrap the service error in a StorageException. 
> 179                 // no op 
> 180             } 
> 181 
 
> 182             // Just wrap in StorageException 
> 183             return new StorageException(reqResult, ex.Message, ex); 
> 184         } 

This is a good example of a place where catch (Exception) can be really painful. (We spent multiple days debugging a problem that would have been immediately obvious if this exception had not been swallowed.) If calling CoreTranslate results in an AssemblyLoadException, that exception should be allowed to propogate up the stack. StorageException can be recoverable - if you expect a blob might not exist, checking for that is common. But if the Storage SDK silently returns a different StorageException when a required assembly is not present, the caller can't see what the real problem is or understand why it's getting an incomplete StorageException even though the HTTP response had full errorcode/errormessage data.

**Comment by pemari-msft** [comment]
Sorry for the delay. Yes, this was a purposeful fix to prevent an hang if this dependency was missing. Glad you got it figured out!

**Comment by davidmatson** [comment]
Please don't close this bug. The bug is that we're swallowing real exceptions and returning incomplete data in that scenario. This catch all block cost us two days.

**Comment by pemari-msft** [comment]
The alternative causes a hang in some cases, which is unfortunately an untenable situation as well. This issue and the solution we implemented was also customer reported and requested, so I'm not sure there's a good answer here. I do apologize for your inconvenience over this.

**Comment by davidmatson** [comment]
Can you link to the hang case? Why does allowing an AssemblyLoadException to propogate cause a hang?

The title of the bug still stands - the body contains extended error information, but the SDK's exception is missing it. At a minimum, could we take out the code "that's marked should never get here" that triggers this assembly load that ultimately results in a null ExtendedErrorInformation?

In other words, I think there is a solution in this case that results in non-null ExtendedErrorInformation - it's present in the HTTP response.

**Comment by centur** [comment]
Hi guys. Just curious on this bug progress - I can observe this problem in latest 7.2 library - `StorageExtendedErrorInformation ExtendedErrorInformation` property is not populated correctly for certain errors.

**Comment by mirobers** [comment]
Currently, this is expected behavior when DLL dependencies are missing. We are tracking this as an improvement request to support the scenario where some dependencies can be safely omitted for some workflows. For now, we expect the library to ship with all of its dependencies.

If you are seeing omitted extended error information even with all DLL dependencies present, then you are seeing something other than what is described in this issue.

**Comment by centur** [comment]
yes, it's because of missing `Microsoft.Data.Services.Client.dll` but why the latest version doesn't make this dependency explicit\mandatory ?  

I'd consider it as a bug, not as an improvement - this library is very important for functioning of many other Azure-related libraries. And having a different runtime behaviour for different environments is a serious problem.
 It's clearly a bug because noone can safely write any code to check, say for transience of storage exception. Simple function must always do a double job:
```
                private bool isTransientException(Exception exc)
		{
			var storageException = exc as StorageException;
			if (storageException == null) return false;
			
			// This is the ugly case for missing Microsoft.Data.Services.Client.dll
			if (storageException.RequestInformation.HttpStatusMessage == "The specified container does not exist.") return true;
			// This is the case when Microsoft.Data.Services.Client.dll
			if (storageException.RequestInformation.ExtendedErrorInformation?.ErrorCode == BlobErrorCodeStrings.BlobNotFound) return true;

			// by default all errors are non transient
			return false;
		}
```

If one would rely only on ExtendedErrorInformation - this check would not work properly for missing Client.dll. And this is depends on runtime... We can't write reliable code with such approach...

**Comment by pemari-msft** [comment]
It is an explicit dependency. I'm afraid I don't quite understand your point.

**Comment by centur** [comment]
My understanding of explicit dependencies - library must not function without satisfying all explicit dependencies. If one of the explicit dependencies is missing - library can't work partially. This breaks the library contract, which also includes exceptions and information about exceptions. 

So this library says - my client\consumer can use any details in `StorageException.RequestInformation,ExtendedErrorInformation` to handle non-successful response from WinAzureStorage Api.
This exception is public as well as nested details\properties - thus it's a cross-domain object, so sender (library) must populate all possible data there.

Current behaviour of the library doesn't guarantee that exception received in response to the same API error (e.g. 409 Conflict) on one server (Server1 with M.D.S.Client.dll referenced\re-bound) will have all the same properties as the exception received on another server (Server 2 without M.D.S.Client.dll referenced\binded properly). This means that depending on the server context - consumer must have 2 different checks to handle 2 different exceptions ( see the snippet above). 

As I can see - nuget with Microsoft.Data.Services.Client.dll is being installed as a dependency for WAS nuget, but the issue is that any code without Microsoft.Data.Services.Client.dll would still work, but IMHO it must break immediately if the necessary MDS.Client.dll is missing.

**Comment by mirobers** [comment]
What I think you're asking for is an assembly load exception rather than silently failing to populate extended error info. This is in effect the work we are tracking, except tweaked to make the breaking change significantly less severe for users who happen to be running without some dependencies today.

The general case of detecting a bad environment is trickier. In .NET, dependencies are loaded dynamically as needed so it's not practical to fail fast in the presence of a corrupt environment. You will get apparently working behavior until the DLL load is attempted. Also note that environment corruption is not something a library can fully handle itself. You will need to ensure integrity of your deployment.

I am fine with tracking it as a bug. I'm reopening this issue for tracking.

**Comment by centur** [comment]
@mirobers I understand that, env corruption is an example to illustrate the problem. One may not even know about env corruption because of typical "it works on my machine" and "...it works on dev environment too". Fail fast here is the best way to go, IMO, but I understand the pain of breaking changes especially when it breaks on prod environment. 
Leaving an option for users to update the nuget but still run in "apparently working" env without some dependencies - up to you, I'd rather prefer to have a hard breaking change for major version upgrade.

They are only in "semi-safe" state in less severe case. If Lib can signal somehow about this problem, at least a nagging warning on build- it would be great.

Thanks for re-opening the bug.

**Comment by mathewc** [comment]
this also affects us in the WebJobs SDK. We rely on ExtendedErrorInfo for lots of decisions, and we've had customers report hard to diagnose issues whose underlying cause was tracked back to this issue.

We're putting in a temporary check to fail fast: https://github.com/Azure/azure-webjobs-sdk/pull/938, until this is fixed.

**Comment by pemari-msft** [comment]
We fixed this in version 8.0 of our library. Now the behavior doesn't happen in blob/file/queue services and we throw if the dependency is missing where it's truly needed in table service.

**Comment by davidmatson** [comment]
Great - thanks!
